lib/auth/trustedcluster.go

```diff
@@ -69,9 +69,24 @@ func (a *Server) UpsertTrustedCluster(ctx context.Context, trustedCluster types.
 	logger := log.WithField("trusted_cluster", trustedCluster.GetName())
 
 	// change state
+	if err := a.checkLocalRoles(ctx, trustedCluster.GetRoleMap()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Update role map
+	if existingCluster != nil && !cmp.Equal(existingCluster.GetRoleMap(), trustedCluster.GetRoleMap()) {
+		if err := a.UpdateUserCARoleMap(ctx, existingCluster.GetName(), trustedCluster.GetRoleMap(),
```

Comment:
So we've updated the role map in the CA here. What happens if anything below fails? Should we restore the original role map? Otherwise we risk the situation where it's updated in the user CA but not updated in the trusted cluster, no?

Reply:
I've added a solution using named returns and modifying the userca in a  defer if the err is nil.